probe-rs/src/architecture/arm/sequences/infineon.rs

```diff
@@ -0,0 +1,249 @@
+
+use super::ArmDebugSequence;
+
+/// An Infineon XMC4xxx MCU.
+pub struct XMC4000 {
+    halt_after_reset_state: Mutex<Option<(bool, u32)>>,
+}
+
+impl XMC4000 {
+    /// Create the sequencer for an Infineon XMC4000.
+    pub fn create() -> Arc<dyn ArmDebugSequence> {
+        Arc::new(Self {
+            halt_after_reset_state: Mutex::new(None),
+        })
+    }
+}
+
+impl ArmDebugSequence for XMC4000 {
+    // We have a weird halt-after-reset sequence. It's described only in prose, not in a CMSIS pack
+    // sequence. Per XMC4700/XMC4800 reference manual v1.3 § 28-8:
+    //
+    // > A Halt after system reset (Warm Reset) can be achieved by programming a break point at the
+    // > first instruction of the application code. Before a Warm Reset the CTRL/STAT.CDBGPWRUPREQ
+    // > and the DHCSR.C_DEBUGEN setting has to be ensured. After a system reset, the HAR situation
+    // > is not considered, as the reset is not coming from "Power-on Reset".
+    // >
+    // > Note: The CTRL/STAT.CDBGPWRUPREQ and DHCSR.C_DEBUGEN does not have to be set after a
+    // > system reset, if they have already been set before.
+    // >
+    // > A tool hot plug condition allows to debug the system starting with a tool registration
+    // > (setting CTRL/STAT.CDBGPWRUPREQ register) and a debug system enable (setting the
+    // > DHCSR.C_DEBUGEN register). Afterwards break points can be set or the CPU can directly by
+    // > HALTED by an enable of C_HALT.
+    //
+    // So:
+    // * ResetCatchSet must determine the first user instruction and set a breakpoint there.
+    // * ResetCatchClear must restore the clobbered breakpoint, if any.
+
+    fn reset_catch_set(
+        &self,
+        core: &mut Memory,
+        _core_type: probe_rs_target::CoreType,
+        _debug_base: Option<u64>,
+    ) -> Result<(), Error> {
+        tracing::trace!("performing XMC4000 ResetCatchSet");
+
+        // We need to find the "first instruction of the application code". The XMC4xxx system
+        // software ("SSW") has a vector table at the start of boot ROM. XMC4700/XMC4800 reference
+        // manual v1.3 § 2-31:
+        //
+        // > On system reset, the vector table is fixed at address 0x00000000.
+        //
+        // That reset vector runs first, but it doesn't qualify as "application code". § 27.2:
+        //
+        // > This section describes the startup sequence of the XMC4[78]00 as a process taking place
+        // > before user application software takes control of the system.
+        //
+        // So: where does the application begin? That depends on the boot mode! The boot mode
+        // depends on the System Control Unit (SCU) Startup Control (STCON) register, which includes
+        // both a software-settable field and a power-on-resettable field. We want a normal boot,
+        // because normal boots are sane and reasonable.
+        let stcon = core.read_word_32(0x50004010)?;
+        if stcon & 0x3 != 0 {
```

Comment:
Would it be possible to add this STCON register as a struct with bitfield impl instead of keeping it a reaw u32? It makes usage in code and logging nicer :)

Reply:
Yes! Good call.

Reply:
I made it a bitfield in https://github.com/probe-rs/probe-rs/pull/1301/commits/f8e112518c9ddb460f9a99d7916606c3148c40ee.